insolar/bus/retryer.go

```diff
@@ -29,53 +29,75 @@ import (
 	"github.com/pkg/errors"
 )
 
-// SendWithRetry sends message to specified role, using provided Sender.SendRole. If error with CodeFlowCanceled
-// was received, it retries request after pulse on current node will be changed.
-// Replies will be written to the returned channel. Always read from the channel using multiple assignment
-// (rep, ok := <-ch) because the channel will be closed on timeout.
-func SendRoleWithRetry(
-	ctx context.Context, b Sender, a pulse.Accessor, msg *message.Message, role insolar.DynamicRole, object insolar.Reference, tries uint,
-) (<-chan *message.Message, func()) {
-	r := newRetryer(b, a)
-	go r.send(ctx, msg, role, object, tries)
-	return r.replyChan, r.clientDone
-}
-
-type retryer struct {
+type RetrySender struct {
 	sender        Sender
 	pulseAccessor pulse.Accessor
 
 	tries uint
 
-	once              sync.Once
-	done              chan struct{}
-	processingStarted sync.Mutex
-	replyChan         chan *message.Message
-	isReplyChanClosed bool
+	once      sync.Once
+	done      chan struct{}
+	replyChan chan *message.Message
 }
 
-func newRetryer(sender Sender, pulseAccessor pulse.Accessor) *retryer {
-	r := &retryer{
+func NewRetrySender(sender Sender, pulseAccessor pulse.Accessor, tries uint) *RetrySender {
+	r := &RetrySender{
 		sender:        sender,
 		pulseAccessor: pulseAccessor,
+		tries:         tries,
 		done:          make(chan struct{}),
 		replyChan:     make(chan *message.Message),
 	}
 	return r
 }
 
-func (r *retryer) clientDone() {
+// SendWithRetry sends message to specified role, using provided Sender.SendRole. If error with CodeFlowCanceled
+// was received, it retries request after pulse on current node will be changed.
+// Replies will be written to the returned channel. Always read from the channel using multiple assignment
+// (rep, ok := <-ch) because the channel will be closed on timeout.
+func (r *RetrySender) SendRole(
+	ctx context.Context, msg *message.Message, role insolar.DynamicRole, ref insolar.Reference,
+) (<-chan *message.Message, func()) {
+	go func() {
+		defer close(r.replyChan)
+		logger := inslogger.FromContext(ctx)
+		var lastPulse insolar.PulseNumber
+
+		select {
+		case <-r.done:
+			return
+		default:
+		}
+
+		received := false
+		for r.tries > 0 && !received {
+			var err error
+			lastPulse, err = r.waitForPulseChange(ctx, lastPulse)
+			if err != nil {
+				logger.Error(errors.Wrap(err, "can't wait for pulse change"))
+				break
+			}
+
+			reps, done := r.sender.SendRole(ctx, msg, role, ref)
+			received = r.tryReceive(ctx, reps)
+			r.tries--
+			done()
+		}
+
+		if r.tries == 0 {
+			logger.Error(errors.Errorf("flow cancelled, retries exceeded"))
```

Comment:
if on last try we got a result `received`, this line still would be lead to an error message.

Reply:
Good point! Changed condition